subsys/bluetooth/services/ans.c

```diff
@@ -0,0 +1,496 @@
+	((IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_SIMPLE_ALERT) << BT_ANS_CAT_SIMPLE_ALERT) |           \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_EMAIL) << BT_ANS_CAT_EMAIL) |                         \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_NEWS) << BT_ANS_CAT_NEWS) |                           \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_CALL) << BT_ANS_CAT_CALL) |                           \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_MISSED_CALL) << BT_ANS_CAT_MISSED_CALL) |             \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_SMS_MMS) << BT_ANS_CAT_SMS_MMS) |                     \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_VOICE_MAIL) << BT_ANS_CAT_VOICE_MAIL) |               \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_SCHEDULE) << BT_ANS_CAT_SCHEDULE) |                   \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_HIGH_PRI_ALERT) << BT_ANS_CAT_HIGH_PRI_ALERT) |       \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_INSTANT_MESSAGE) << BT_ANS_CAT_INSTANT_MESSAGE))
+
+/* Build time ANS supported category bit mask */
+#define BT_ANS_UNALRT_CAT_MASK                                                                     \
+	((IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_SIMPLE_ALERT) << BT_ANS_CAT_SIMPLE_ALERT) |          \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_EMAIL) << BT_ANS_CAT_EMAIL) |                        \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_NEWS) << BT_ANS_CAT_NEWS) |                          \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_CALL) << BT_ANS_CAT_CALL) |                          \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_MISSED_CALL) << BT_ANS_CAT_MISSED_CALL) |            \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_SMS_MMS) << BT_ANS_CAT_SMS_MMS) |                    \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_VOICE_MAIL) << BT_ANS_CAT_VOICE_MAIL) |              \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_SCHEDULE) << BT_ANS_CAT_SCHEDULE) |                  \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_HIGH_PRI_ALERT) << BT_ANS_CAT_HIGH_PRI_ALERT) |      \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_INSTANT_MESSAGE) << BT_ANS_CAT_INSTANT_MESSAGE))
+
+/* As per spec, ensure at least one New Alert category is supported */
+BUILD_ASSERT(BT_ANS_NALRT_CAT_MASK != 0,
+	     "At least one ANS New Alert category must be enabled in Kconfig");
+
+static uint16_t alert_sup_cat_bit_mask = BT_ANS_NALRT_CAT_MASK;
+static uint16_t unread_sup_cat_bit_mask = BT_ANS_UNALRT_CAT_MASK;
+
+/* Command ID definitions */
+#define BT_ANS_SEND_ALL_CATEGORY 0xff
+enum bt_ans_command_id {
+	BT_ANS_CTRL_ENABLE_NEW_ALERT,
+	BT_ANS_CTRL_ENABLE_UNREAD,
+	BT_ANS_CTRL_DISABLE_NEW_ALERT,
+	BT_ANS_CTRL_DISABLE_UNREAD,
+	BT_ANS_CTRL_NOTIFY_NEW_ALERT_IMMEDIATE,
+	BT_ANS_CTRL_NOTIFY_UNREAD_IMMEDIATE,
+};
+
+/* Struct definition for Alert Notification Control Point */
+struct alert_ctrl_p {
+	uint8_t cmd_id;
+	uint8_t category;
+} __packed;
+
+/* Struct definition for New Alert */
+struct new_alert {
+	uint8_t category_id;
+	uint8_t num_new_alerts;
+	char text_string[BT_ANS_MAX_TEXT_STR_SIZE];
```

Comment:
```suggestion
	char text_string[BT_ANS_MAX_TEXT_STR_SIZE + 1];
```

Don't we need to 1 additional octet here for the NULL terminator?

Reply:
I don't think so since the spec talks about this string in terms of octets. So I would assume the NULL terminator is part of the 'max size' already.